docs/design/README.md

```diff
@@ -0,0 +1,130 @@
+to source-controller api to fetch repo files instead of reaching out to git directly. 
+
+
+***Note:*** *In future, kraan will support packaging addons in formats other 
+than helm charts e.g kustomization*
+ 
+### Spec
+The addons are first packaged as HelmRelease custom resources which represents
+a helm chart. A sample helm release custom resource is shown below.
+For more details on *HelmRelease*, check
+[here](https://docs.fluxcd.io/projects/helm-operator/en/stable/helmrelease-guide/introduction/).
+
+
+```yaml
+# podinfo.yaml
+---
+apiVersion: helm.fluxcd.io/v1
+kind: HelmRelease
+metadata:
+  name: podinfo
+  namespace: default
+spec:
+  chart:
+    repository: https://stefanprodan.github.io/podinfo
+    name: podinfo
+    version: 3.2.0
+```
+
+###### Example
+A sample addon layer custom resource is shown below. 
+```yaml
+apiVersion: kraan.io/v1alpha1
+kind: AddonsLayer
+metadata:
+  name: bootstrap
+
+spec:
+  version: 1.0.0
+  hold: true
+  interval: 1m
+  source:
+    name: addons
+    namespace: gitops-system
+    path: ./addons/bootstrap
+  prereqs:
+      k8sVersion: "1.16"
+---
+apiVersion: kraan.io/v1alpha1
+kind: AddonsLayer
+metadata:
+  name: base
+
+spec:
+  version: 1.0.1
+  interval: 1m
+  source: 
+    name: addons
+    namespace: gitops-system
+    path: ./addons/base
+  prereqs:
+      k8sVersion: "1.16"
+      dependsOn:
+        - bootstrap@1.0.0
+```
+
+## Reconciliation Logic
+Kraan controller is built using the [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) project.
+From a controllers' perspective, reconciliation is where it attempts to bring the
+current state of the layer into its desired state. The reconciliation logic is 
+kicked off whenever there is a change detected in the AddonLayer custom resource.
+It could be an addition, deletion or an update to the resource.Following is 
+the step by step reconciliation logic in kraan. Whenever an event is 
+received for an AddonLayer,
+
+* check whether prereqs.k8sVersion == cluster version. If the cluster is not
+  yet in the expected version, return back without throwing an error. 
+  Controllers' periodic sync will trigger the reconcile next time.
+* execute a custom prune logic, where any addon that is present in the cluster
+  but NOT in the remove the layer from the cluster. Basically any addon that is 
+  not part of the current layer spec is removed from the cluster. Note that removal
+  of an addon does not check for dependencies between layers. If the prune is
+  not successful, then return an error, which should trigger an exponential
+  back off.
+* Check if the dependent layers are succesfully deployed. This is a multi step process.
+    * The dependent layer is identified by a combination of name and a version.
+      i.e base@0.1.2 means that a layer with name "base" and spec.version = 0.1.2 is expected.
+    * If the dependent layer isn't found, return nil and requeue the request
+      to process after 10 mins (RequeueAfter: 10mins).
```

Comment:
I suggest we don't need golang code in this text, remove  (RequeueAfter: 10mins)